### PR TITLE
Allows to update permissions without payload

### PR DESCRIPTION
**Pull request by gepd**
## What does this PR do?

Currently, to update the permission of a document, you also have to update some kind of data, or at least pass some kind of data.

This does not work:

```js
await database.updateDocument(listCollection, list, {}, [`user:${user.$id}`]);
```

So to update permissions we have to pass some data, even if the data is the same as the current one:

```js
await database.updateDocument(listCollection, list, {name: "test"}, [`user:${user.$id}`]);
```

This change will allow to send an empty payload, but empty payload and read/write permission will keep throwing an error

```js
await database.updateDocument(listCollection, list, {}, [], []);
```

## Test Plan

No new test

## Related PRs and Issues

https://github.com/appwrite/appwrite/issues/2850

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

yes

### CODE REVIEW

**Comment by gepd** [comment]
The failing test doesn't seems to be related with my changes

**Comment by TorstenDittmann** [comment]
Restarted the tests 👍🏻 

I think it would make sense to make the `data` param optional. Otherwise it would be required but optional 🙂

**Comment by gepd** [comment]
@TorstenDittmann good catch I totally forgot about that, I'm not good at types in PHP, but reading the manual, that should be the way to make it optional

let me know if that is enough 😀

**Comment by TorstenDittmann** [comment]
Utopia has all those `->param` before the callback. If the last argument is true, the parameter of the rest API is marked as optional 😉

**Comment by gepd** [comment]
@TorstenDittmann I think I got it, Is everyting good now?

File: app/controllers/api/database.php [hunk 1]
```diff
@@ -1868,7 +1868,7 @@ function createAttribute(string $collectionId, Document $attribute, Response $re
     ->inject('usage')
     ->inject('events')
     ->inject('mode')
-    ->action(function (string $collectionId, string $documentId, string|array $data, ?array $read, ?array $write, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events, string $mode) {
+    ->action(function (string $collectionId, string $documentId, string|array|null $data, ?array $read, ?array $write, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events, string $mode) {
```
Comment (everly-gif) [inline app/controllers/api/database.php:1871]:
```suggestion
    ->action(function (string $collectionId, string $documentId, string|array $data, ?array $read, ?array $write, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events, string $mode) {
```
you can remove `null` from the types here , now that you have set the the last argument to true for the param.

**Review by Meldiron — CHANGES_REQUESTED** [review]

**Comment by gepd** [comment]
@Meldiron can you reset the tests?

**Comment by gepd** [comment]
I need help
After checking the test I realized there is something else looking for the document structure

I'm getting error `400` which I believe comes from here: https://github.com/gepd/appwrite/blob/feat-empty-payload/app/controllers/api/database.php#L1943

I'm not sure but, is this related to utopia php?

Seems like here is looking for a specific structure: https://github.com/utopia-php/database/blob/main/src/Database/Database.php#L1169-L1173

@Meldiron @TorstenDittmann

File: app/controllers/api/database.php [hunk 2]
```diff
@@ -1900,8 +1900,8 @@ function createAttribute(string $collectionId, Document $attribute, Response $re
 
         $data = (\is_string($data)) ? \json_decode($data, true) : $data; // Cast to JSON array
 
-        if (empty($data)) {
-            throw new Exception('Missing payload', 400, Exception::DOCUMENT_MISSING_PAYLOAD);
+        if (empty($data) && empty($read) && empty($write)) {
```
Comment (Meldiron) [inline app/controllers/api/database.php:1903]: I am unsure about the `empty` check here. Would this allow me to set permissions to an empty array - remove all current permissions?

If not, we might need check with `isset` or `=== null` or `=== ''` or change default value when defining `->param()` to something else. Not sure, please play around to see what works the best.

**Comment by gepd** [comment]
I have finally fixed, I have updated the test, I think not deleting the collection was causing problems

@Meldiron @TorstenDittmann

**Review by Meldiron — APPROVED** [review]
Looks good to me

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**